source/common/http/filter_manager.cc

```diff
@@ -1044,9 +1052,45 @@ void FilterManager::encodeHeaders(ActiveStreamEncoderFilter* filter, ResponseHea
     }
   }
 
-  const bool modified_end_stream = (end_stream && continue_data_entry == encoder_filters_.end());
+  // See if the response would be written by local_reply_.
+  if (!did_rewrite_) {
+    do_rewrite_ =
+        local_reply_.match(filter_manager_callbacks_.requestHeaders().ptr(),
+                           *filter_manager_callbacks_.responseHeaders(),
+                           filter_manager_callbacks_.responseTrailers().ptr(), stream_info_);
+  }
+
+  bool modified_end_stream = (end_stream && continue_data_entry == encoder_filters_.end());
+  const bool original_modified_end_stream = modified_end_stream;
+  ENVOY_STREAM_LOG(
+      debug, "FilterManager::encodeHeaders: end_stream={}, modified_end_stream={}, do_rewrite_={}",
+      *this, end_stream, modified_end_stream, do_rewrite_);
+  if (do_rewrite_) {
+    // _This_ actually sets buffered_response_data_ internally.
+    rewriteResponse();
+
+    if (buffered_response_data_->length() > 0) {
+      // If we're going to rewrite the response here, then modified_end_stream can no longer be true
+      // because we have a body now.
+      modified_end_stream = false;
+    }
+  }
+
   state_.non_100_response_headers_encoded_ = true;
   filter_manager_callbacks_.encodeHeaders(headers, modified_end_stream);
+
+  // Encode the rewritten response right away if the original `modified_end_stream` was true.
+  // If it wasn't, then we know a body will be encoded later, and we'll let that function
+  // take care of encoding the rewritten response.
+  if (do_rewrite_ && original_modified_end_stream && buffered_response_data_->length() > 0) {
```

Comment:
I think the constant checking for buffered_response_data->length() > 0 would be improved if the local_reply_.rewrite interface provided a way of returning a null optional for the rewritten body, in the case where there is no configured body rewrite.